plugin/pulpcore/plugin/stages/association_stages.py

```diff
@@ -0,0 +1,150 @@
+        """
+        with ProgressBar(message='Associating Content') as pb:
+            batch = []
+            shutdown = False
+            while True:
+                try:
+                    declarative_content = in_q.get_nowait()
+                except asyncio.QueueEmpty:
+                    if not batch and not shutdown:
+                        declarative_content = await in_q.get()
+                        batch.append(declarative_content)
+                        continue
+                else:
+                    batch.append(declarative_content)
+                    continue
+
+                content_q_by_type = defaultdict(lambda: Q(pk=None))
+                for declarative_content in batch:
+                    if declarative_content is None:
+                        shutdown = True
+                        continue
+                    try:
+                        unit_key = declarative_content.content.natural_key()
+                        self.unit_keys_by_type[type(declarative_content.content)].remove(unit_key)
+                    except KeyError:
+                        model_type = type(declarative_content.content)
+                        unit_key_dict = declarative_content.content.natural_key_dict()
+                        unit_q = Q(**unit_key_dict)
+                        content_q_by_type[model_type] = content_q_by_type[model_type] | unit_q
+
+                for model_type, q_object in content_q_by_type.items():
+                    queryset = model_type.objects.filter(q_object)
+                    self.new_version.add_content(queryset)
+                    pb.done = pb.done + queryset.count()
+                    pb.save()
+
+                if shutdown:
+                    break
+                batch = []
+
+            for unit_type, ids in self.unit_keys_by_type.items():
+                if ids:
+                    units_to_unassociate = Q()
+                    for unit_key in self.unit_keys_by_type[unit_type]:
+                        query_dict = {}
+                        for i, key_name in enumerate(unit_type.natural_key_fields()):
+                            query_dict[key_name] = unit_key[i]
+                        units_to_unassociate |= Q(**query_dict)
+                    await out_q.put(unit_type.objects.filter(units_to_unassociate))
+            await out_q.put(None)
+
+
+class ContentUnitUnassociation(BaseStage):
+    """
+    A Stages API stage that unassociates content units from `new_version`.
+
+    This stage creates a ProgressBar named 'Un-Associating Content' that counts the number of units
+    un-associated. Since it's a stream the total count isn't known until it's finished.
+
+    Args:
+        new_version (:class:`~pulpcore.plugin.models.RepositoryVersion`): The repo version this
+            stage unassociates content from.
+    """
+
+    def __init__(self, new_version, *args, **kwargs):
+        self.new_version = new_version
+        super().__init__(*args, **kwargs)
```

Comment:
Same re: calling *super* as: [here](https://github.com/pulp/pulp/pull/3559/files#r207544584)
Same re: *args, kwargs* as: [here](https://github.com/pulp/pulp/pull/3559/files#r207543875)

Reply:
yup will fix in next push.